influxdb3_clap_blocks/src/datafusion.rs

```diff
@@ -0,0 +1,125 @@
+use std::collections::HashMap;
+
+use datafusion::config::ConfigExtension;
+use iox_query::config::IoxConfigExt;
+
+/// Extends the standard [`HashMap`] based DataFusion config option in the CLI with specific
+/// options (along with defaults) for InfluxDB 3 OSS/Pro. This is intended for customization of
+/// options that are defined in the `iox_query` crate, e.g., those defined in [`IoxConfigExt`]
+/// that are relevant to the monolithinc versions of InfluxDB 3.
+#[derive(Debug, clap::Parser, Clone)]
+pub struct IoxQueryDatafusionConfig {
+    /// When multiple parquet files are required in a sorted way (e.g. for de-duplication), we have
+    /// two options:
+    ///
+    /// 1. **In-mem sorting:** Put them into `datafusion.target_partitions` DataFusion partitions.
+    ///    This limits the fan-out, but requires that we potentially chain multiple parquet files into
+    ///    a single DataFusion partition. Since chaining sorted data does NOT automatically result in
+    ///    sorted data (e.g. AB-AB is not sorted), we need to preform an in-memory sort using
+    ///    `SortExec` afterwards. This is expensive.
+    /// 2. **Fan-out:** Instead of chaining files within DataFusion partitions, we can accept a
+    ///    fan-out beyond `target_partitions`. This prevents in-memory sorting but may result in OOMs
+    ///    (out-of-memory) if the fan-out is too large.
+    ///
+    /// We try to pick option 2 up to a certain number of files, which is configured by this
+    /// setting.
+    #[clap(
+        long = "datafusion-max-parquet-fanout",
+        env = "INFLUXDB3_DATAFUSION_MAX_PARQUET_FANOUT",
+        default_value = "1000",
```

Comment:
Good to call this out. I copied the comment from IOx/core to preserve the context it provided. I think we may need to tune this a bit, or it could be possible to base the default on the system memory, and how we allocate memory in different modes in pro.

As it stands, with the low default of 40, we are getting OOMs with the fallback, i.e., non-fanout, query plan, so we should know soon if increasing this much makes the problem worse or not. Based on https://github.com/influxdata/influxdb_pro/issues/308#issuecomment-2562955195, this default may be a bit low/out-dated (perhaps the way the DataFusion plan handles fanout is different than when the default was decided). There are some distributed clusters in IOx setting this to 800 as per https://github.com/influxdata/influxdb_pro/issues/308#issuecomment-2563245404.

We'll see how this goes - at the minimum, I got the env vars switched from `INFLUXDB_IOX_` to `INFLUXDB3_` 😄

Reply:
I might have misunderstood the docs for this setting, I interpreted it as, the higher this number the more files it tries to fan-out, which leads to OOMs. If we don't fan-out then it leads to doing expensive in memory sorting (guessing without running into OOMs?).

Reply:
> (guessing without running into OOMs?)

Unfortunately, though, it _is_ OOM'ing without the fanout, while _not_ OOM'ing with the fanout, so we may need to update this doc comment.

Reply:
I think the memory sort is going to OOM, unless you set a memory limit on DF, but in that case it just means that the query will get killed and return a resource exhaustion error. The only way around that I can think of is if spill to disk is enabled, but that's not really much better either.

I think the fanout setting should effectively be ignored (i.e. set to whatever the max of the type is). Resorting the data is always going to be more expensive and completely unnecessary in our case.

If DF allocates an arrow buffer for each input file, then you'd have that size * num of files. The Arrow buffer could be quite large if there are very wide tables and depending on the size of that buffer. I think one way to counter this would be to make sure that the pre-allocated buffer is limited in size or scaled down depending on the number of input files.